rslib/src/scheduler/states/load_balancer.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::cmp::Ordering;
+use std::collections::HashSet;
+
+use crate::decks::DeckId;
+use crate::notes::NoteId;
+use crate::storage::SqliteStorage;
+
+const MAX_LOAD_BALANCE_INTERVAL: u32 = 90;
+const PERCENT_BEFORE: f32 = 0.1;
+const PERCENT_AFTER: f32 = 0.1;
+const DAYS_MIN_BEFORE: i32 = 1;
+const DAYS_MIN_AFTER: i32 = 1;
+const DAYS_MAX_BEFORE: i32 = 6;
+const DAYS_MAX_AFTER: i32 = 4;
+
+pub struct LoadBalancer<'a> {
+    today: u32,
+    note_id: NoteId,
+    deck_id: Option<DeckId>,
+    avoid_siblings: bool,
+    storage: &'a SqliteStorage,
+}
+
+impl<'a> LoadBalancer<'a> {
+    pub fn new_from_collection(
+        today: u32,
+        storage: &'a SqliteStorage,
+        note_id: NoteId,
+        avoid_siblings: bool,
+    ) -> LoadBalancer<'a> {
+        LoadBalancer {
+            today,
+            note_id,
+            avoid_siblings,
+            storage,
+            deck_id: None,
+        }
+    }
+
+    pub fn new_from_deck(
+        today: u32,
+        storage: &'a SqliteStorage,
+        note_id: NoteId,
+        deck_id: DeckId,
+        avoid_siblings: bool,
+    ) -> LoadBalancer<'a> {
+        LoadBalancer {
+            today,
+            note_id,
+            avoid_siblings,
+            storage,
+            deck_id: Some(deck_id),
+        }
+    }
+
+    pub fn find_interval(&self, interval: u32) -> u32 {
+        // if we're sending a card far out into the future, the need to balance is low
+        if interval > MAX_LOAD_BALANCE_INTERVAL {
+            return interval;
+        }
+
+        // determine the range of days to check
+        let before_range =
+            ((interval as f32 * PERCENT_BEFORE) as i32).clamp(DAYS_MIN_BEFORE, DAYS_MAX_BEFORE);
+        let after_range =
+            ((interval as f32 * PERCENT_AFTER) as i32).clamp(DAYS_MIN_AFTER, DAYS_MAX_AFTER);
```

Comment:
It doesn't respect the fuzz range. The fuzz range is not simple +- 10%.

https://github.com/open-spaced-repetition/fsrs4anki-helper/blob/18e9ec348a514d8499e252a9c5233757193622a3/utils.py#L113-L146

Reply:
The fuzz function in Anki:

https://github.com/ankitects/anki/blob/33a923797afc9655c3b4f79847e1705a1f998d03/rslib/src/scheduler/states/fuzz.rs#L103-L115

Reply:
I have no idea how I completely missed this function, I was all over that file looking for something like this.

Reply:
updated to use `fuzz_bounds`